File: R/mriqc.R
list_jobs_whole_mriqc <- list_jobs_status_fmriprep

list_jobs_status_mriqc <- function(check_file_sum = FALSE) {
  # 8 files are generated for each session
  num_files_ses <- 8L
  col_names_chk <- c("subject", "session")
  files_mriqc <- fs::path(path_derivatives, "mriqc") |>
    fs::dir_ls(type = "file", regexp = "sub") |>
    str_match("sub-([:alnum:]+)_ses-(\\d)") |>
    as_tibble(.name_repair = ~ c("whole", col_names_chk)) |>
    count(pick(all_of(col_names_chk)))
  if (check_file_sum) {
    files_mriqc |>
      mutate(
        status = case_when(
          n == num_files_ses ~ "done",
          n == 0 ~ "todo",
          .default = "incomplete"
        )
      ) |>
      select(all_of(col_names_chk), status)
  } else {
    files_mriqc |>
      mutate(
        status = if_else(
          n >= 1,
          "done", "todo"
        )
      ) |>
      select(all_of(col_names_chk), status)
  }
}

commit_mriqc <- function(sublist, file_sublist = NULL, ...) {
  rlang::check_dots_empty()
  if (is.null(file_sublist)) {
    dir_file_sublist <- fs::path(path_tmp, "qsub", "mriqc")
    if (!fs::dir_exists(dir_file_sublist)) {
      fs::dir_create(dir_file_sublist)
    }
    file_sublist <- fs::path(
      dir_file_sublist,
      format(now(), "sublist-%Y%m%d_%H%M%S")
    )
  }
  sublist |>
    select(subject, session) |>
    write_delim(file_sublist, col_names = FALSE)
  num_jobs <- nrow(sublist)
  fs::path(path_qsub, "mriqc.tmpl.qsub") |>
    read_file() |>
    str_glue() |>
    commit(
      "mriqc",
      num_jobs = num_jobs,
      file_sublist = file_sublist
    )
  invisible()
}

File: preproc/mriqc.py
from .constants import *
from .utils import *
from pathlib import Path
import pandas as pd
import re


def list_jobs_whole_mriqc():
    # In R, this maps to fmriprep status.
    from .fmriprep import list_jobs_status_fmriprep
    df = list_jobs_status_fmriprep()
    return df[['subject', 'site', 'sid']]


def list_jobs_status_mriqc(check_file_sum=False):
    # collect files under derivatives/mriqc
    base = Path(path_derivatives) / 'mriqc'
    if not base.exists():
        return pd.DataFrame(columns=['subject', 'session', 'status'])
    files = list(base.glob('**/*sub-*'))
    rows = []
    # match sub-..._ses-1 etc
    for f in files:
        m = re.search(r'sub-([A-Za-z0-9]+)_ses-(\d)', f.name)
        if not m:
            continue
        subject = m.group(1)
        session = m.group(2)
        rows.append({'subject': subject, 'session': session})
    df = pd.DataFrame(rows)
    if df.empty:
        return pd.DataFrame(columns=['subject', 'session', 'status'])
    counts = df.groupby(['subject', 'session']).size().reset_index(name='n')
    if check_file_sum:
        counts['status'] = counts['n'].apply(lambda n: 'done' if n == 8 else ('todo' if n == 0 else 'incomplete'))
    else:
        counts['status'] = counts['n'].apply(lambda n: 'done' if n >= 1 else 'todo')
    return counts[['subject', 'session', 'status']]


def commit_mriqc(sublist, file_sublist=None, **kwargs):
    import datetime
    if file_sublist is None:
        dir_file_sublist = Path(path_tmp) / 'qsub' / 'mriqc'
        dir_file_sublist.mkdir(parents=True, exist_ok=True)
        file_sublist = dir_file_sublist / f"sublist-{datetime.datetime.now().strftime('%Y%m%d_%H%M%S')}"
    sublist[['subject', 'session']].to_csv(file_sublist, sep='\t', index=False, header=False)
    num_jobs = len(sublist)
    qsub_fn = Path(path_qsub) / 'mriqc.tmpl.qsub'
    tmpl = qsub_fn.read_text()
    from .utils import render_template
    from .env import get_env_settings
    mapping = get_env_settings()
    mapping.update({'PROJECT_ROOT': str(project_root)})
    cmd = render_template(tmpl, mapping)
    commit(cmd, 'mriqc', num_jobs=num_jobs, file_sublist=file_sublist)
